doc/layer2-forwarding-enhancements/SONiC Layer 2 Forwarding Enhancements HLD.md

```diff
@@ -0,0 +1,444 @@
+
+For VLAN member range creation and deletion, similar logic is added. 
+
+VLAN and port identifier validation are added as part of the logic added in the functions for the new commands.
+
+
+## 3.2 DB Changes
+### 3.2.1 CONFIG DB
+**FDB table for Static MAC**
+
+FDB_TABLE will be set in CONFIG_DB to store static MAC configuration.
+
+**SWITCH table for Aging time configuration**
+
+SWITCH_TABLE will be set in CONFIG_DB to store FDB aging time configuration.
+
+## 3.3 Switch State Service Design
+### 3.3.1 Orchestration Agent
+
+
+**Data Structure changes**
+
+
+The internal FDB data structure  will be enhanced to store the FDB type (static/dynamic). Currently, FDB is stored as c++ set. It will be changed to c++ map to store (key, value), where key is (MAC, bv_id) and value is FDB type.
+FDB type is needed to add Static FDB to saved FDB after flush.
+
+
+	**FDB Data structure in Orchagent:**
+
+
+<img src="images/fdbDataStructure.jpg" width="800" height="500">
+
+In current Sonic implementation, when a FDB event is received with VLAN SAI object ID, the VLAN ID is retrieved by traversing through all the ports and VLANs in the orchagent DB and finding the one that matches the object ID. Also, the Port details are retrieved by traversing all the port and VLANs. This approach is inefficient as it may traverse through the number of VLANs configured for each FDB learn event. Worst case time is O(M*N) where M is the number of MAC events and N is the number of entries in port data structure.
+
+
+A new  unordered_map containing a mapping between SAI object ID (can be either Port object ID or Bridge Port object ID or VLAN object ID) and Name (port or VLAN alias) will be added which can be looked up in O(1) time. The Port or VLAN alias retrieved from this will be looked up in the Port data structure.
+
+
+	**Object ID to alias mapping.**
+
+
+<img src="images/oidToName.jpg" width="800" height="500">
+
+
+**Changes for FDB flush.**
+When a FDB flush request is received, Orchagent will send a bulk flush request to syncd. Individual FDB delete response from SAI will be used to delete FDB entries in ASIC_DB, STATE_DB and orchagent data structure. The delete response will be received as AGED event to syncd, which will delete the entry in ASIC_DB and notify Orchagent.
+
+
+When a port operational status goes down, portsorch will call fdborch API to trigger FDB flush. Only dynamic FDB will be flushed.
+
+
+When a port is removed from VLAN, portsorch will call fdborch API to trigger FDB flush. Both static and dynamic FDB will be flushed. Also, the static FDB will be added to the temporary DB in orchagent when delete response is received, so that it can programmed back when port is added back to VLAN.
```

Comment:
What is the replay strategy of static Mac entries back to vlan? is this temporary db mentioned in orchagent is an in-memory or redis db?  what is the strategy around orchagent crash if it is stored in-memory?

Reply:
It is the existing in memory "saved_fdb_entries". After orchagent crash, SWSS docker restart will   
reconcile from APP_DB.